### PR TITLE
[Stashing] Adds workflow that allows users to take changes with them to new branch

### LINKED ISSUE #7248: [Stashing] Merge conflicts resulting from applying a stash entry does not update state properly
The UI isn't updating correctly when taking changes to the new branch causes a merge conflict and results in:
* the stash dialog not being dismissed even thought the branch was successfully checked out
* files in the _Changes_ view aren't updated (switching to _History_ and back to _Changes_ forces the UI to update itself using the new state and we see that applying worked just fine).
* The branch checkout progress indicator never goes away.

**Repro steps**
1. Branch from `dev
1. Commit change on `feature` 
1. Commit change on `dev` 
1. Make a change on `dev`
1. Switch to `master`
1. Select option to take changes to `master`
1. Click _Switch Branch_

bring_changes_to_new_branch_fails_when_conflicts_are_encountered

### PR DESCRIPTION
**Pull request by iAmWillShepherd**
⚠️ Review #7210 first ⚠️ 

Fixes #7248 - ac1a72f and 7fbce46
Related issues: #6107, #6987 

This PR addresses the following user story:

> I realized after I started doing work that I actually wanted to be doing that work on a different branch. I want to be able to switch branches and retain the uncommitted changes I've made without having to leave GitHub Desktop.

**Important bits**

* Changed `git stash apply <stash>` to `git stash pop <stash>`  after learning that I can specify a stash to pop. `git stash drop <stash>` still exists because we still have to support dropping an arbitrary stash.
* Added a new enum to control what stash related tasks occur before/after a checkout.
* Moved logic for each workflow from the component to `AppStore` to reduce simplify the component and make the relationship between stashing and checking out a branch explicit.

 
**What's not here**

* More robust error-handling. I am letting errors bubble up to the UI during the development so any issues are immediately visible.
* UI for viewing diff of stash

### CODE REVIEW

File: app/src/lib/stores/app-store.ts [hunk 1]
```diff
@@ -2514,6 +2525,22 @@ export class AppStore extends TypedBaseStore<IAppState> {
       return repository
     }
 
+    let shouldPopStash = false
+
+    if (
+      currentBranch !== null &&
```
Comment (outofambit) [inline app/src/lib/stores/app-store.ts]: this is a (non-blocking) nitpick, but could we pull this conditional into an enclosing `if` statement so its easier to follow that either `uncommittedChangesStrategy` will be acted on if `currentBranch !== null`

File: app/src/lib/stores/app-store.ts [hunk 2]
```diff
@@ -4688,6 +4717,21 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
+      branch.name
+    )
+
+    if (stash === null) {
+      log.warn('no stash found that matches')
+      return
+    }
+
+    await popStashEntry(repository, stash.stashSha)
+    log.warn(`Popped stash with commit id ${stash.stashSha}`)
```
Comment (outofambit) [inline app/src/lib/stores/app-store.ts]: why is this `warn` and not `info` or `debug`? doesn't seem like its something we should be even mildly worried about.

File: app/src/ui/stash-and-switch-branch-dialog.tsx [hunk 3]
```diff
@@ -88,24 +89,28 @@ export class StashAndSwitchBranch extends React.Component<
...
   }
 
   private onSubmit = async () => {
-    const {
-      repository,
-      currentBranch,
-      branchToCheckout,
-      dispathcer,
-    } = this.props
+    const { repository, branchToCheckout, dispatcher } = this.props
```
Comment (outofambit) [inline app/src/ui/stash-and-switch-branch-dialog.tsx]: i like this syntax :)

File: app/src/models/uncommitted-changes-strategy.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export enum UncommittedChangesStrategy {
+  askForConfirmation,
```
Comment (outofambit) [inline app/src/models/uncommitted-changes-strategy.ts]: is this the state we use when the popup is first opened and we don't have an answer from the user yet?

File: app/src/lib/stores/app-store.ts (hunk 2 above)
Comment (iAmWillShepherd) [reply app/src/lib/stores/app-store.ts]: Yea, I'll change to `info`. My main concern is that what ever is logged is actually persisted to disk.

File: app/src/models/uncommitted-changes-strategy.ts (hunk 4 above)
Comment (iAmWillShepherd) [reply app/src/models/uncommitted-changes-strategy.ts]: This enum controls what to do about uncommitted changes:
* `askForConfirmation` - informs what to do about changes
* `stashOnCurrentBranch`/`moveToNewBranch` - informs what action to take with changes

**Comment by iAmWillShepherd** [comment]
🍎

File: app/src/lib/stores/app-store.ts [hunk 5]
```diff
@@ -450,7 +455,7 @@ export class AppStore extends TypedBaseStore<IAppState> {
         this.emitUpdate()
       })
       .catch(err => {
-        log.warn(`Unexpected issue when trying to read emoji into memory`, err)
+        log.info(`Unexpected issue when trying to read emoji into memory`, err)
```
Comment (outofambit) [inline app/src/lib/stores/app-store.ts]: i don't think this is the `warn` we wanted to change

Comment (outofambit) [reply app/src/lib/stores/app-store.ts]:
```suggestion
        log.warn(`Unexpected issue when trying to read emoji into memory`, err)
```

File: app/src/lib/stores/app-store.ts [hunk 6]
```diff
@@ -4711,6 +4741,21 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
+      branch.name
+    )
+
+    if (stash === null) {
+      log.warn('no stash found that matches')
+      return
+    }
+
+    await popStashEntry(repository, stash.stashSha)
+    log.warn(`Popped stash with commit id ${stash.stashSha}`)
```
Comment (outofambit) [inline app/src/lib/stores/app-store.ts]: this is still `warn`?
```suggestion
    log.info(`Popped stash with commit id ${stash.stashSha}`)
```

**Pull request merged**